File: allcode/3200-3299/3202maximumLength.py


from leetcode.allcode.competition.mypackage import *

class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [[1] * k for _ in range(n)]  # 以nums[i]结尾，相邻元素和模k为j的序列最大长度
        ans = 1
        for i in range(1, n):
            for t in range(i):
                mod = (nums[i] + nums[t]) % k
                dp[i][mod] = dp[t][mod] + 1
            ans = max(ans, max(dp[i]))
        # print(dp)
        return ans


so = Solution()
print(so.maximumLength(nums = [1,2,3,10,2], k = 6))
print(so.maximumLength(nums = [3,7,6], k = 2))
print(so.maximumLength(nums = [1,2,3,4,5], k = 2))
print(so.maximumLength(nums = [1,4,2,3,1,4], k = 3))

File: c++/src/3000-3999/3202maximumLength.cpp
#include "lc_pub.h"


class Solution {
    public:
    int maximumLength(vector<int>& nums, int k) {
        int n=nums.size();
        for (int i=0;i<n;i++) nums[i]%=k;
        int ans=0;
        for (int i=0;i<k;i++) {
            vector<int> counter(k, 0);  // 结尾是模为k的最长长度
            for (int j=0;j<n;j++) {
                int t=(i+k-nums[j])%k;
                counter[nums[j]]=counter[t]+1;
                ans=max(ans,counter[nums[j]]);
            }
        }
        return ans;
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> nums{1,2,3,4,5};

    Solution so;
    cout << so.maximumLength(nums,2) << endl;
    return 0;
}
